ramips: Cudy WR1300v2 fix mt7613 calibration data length

zekica opened the pull request:
In the current .dts,
offset for calibration data for &pcie1 wifi@0.0 (MT7613) is 0x0000 and
offset for calibration data for &pcie0 wifi@0.0 (MT7603) is 0x8000.

This patch reverses those values to match the actual data:

- data at 0x0000 starts with 0x76 0x03 - start of MT7603 data
- data at 0x8000 starts with 0x76 0x63 - start of MT7663 data

Also, since MT7613 is handled by MT7615 driver, and other devices using MT7615 have reg = <0x8000 0x4da8>; this also needs updating or eeprom data fails to load.

Without this patch, TX power calibration fails on 5GHz and is possibly wrong for 2.4GHz.

najdanovicivan commented:
I've checked this out and it's working properly.

zekica commented:
Additionally, I have confirmed that other devices with MT7613 such as TP-Link Archer C6 v3 has the same 0x4da8 length.

neheb commented:
ping @DragonBluep

DragonBluep commented:
Sorry for the trouble. I read the commit message from the file git history so I got the wrong information. This dts file was renamed from the wr1300(v1) which has MT7612.

The eeprom size looks fine. But are you sure you need to change the offsets? I didn't touch them at all. What's your device revision? V3 doesn't seem to be supported yet. And they are different? https://github.com/openwrt/openwrt/pull/13387

zekica commented:
Thank you for looking into this.

v2 and v3 have the same OpenWrt image [provided by the vendor](http://www.cudytech.com/openwrt_software_download) using the same .dts and build settings. Their .dts correctly lists the offsets - same as this patch. [OpenWrt ToH](https://openwrt.org/toh/hwdata/cudy/cudy_wr1300_v2) also lists v2,v3 as supported.

But more importantly, I have access to one v2 device (@najdanovicivan lent me one) and three v3 devices, and I can confirm that all four work completely fine with this patch. You can also see the calibration data extracted from the devices yourself [in this gist](https://gist.github.com/zekica/efda6cdbc7773154a857d5a370f19d1f) where it is plain to see that 7603 calibration is at offset 0x0000 and 7663 calibration is at offset 0x8000.

My previous attempt at fixing support for v2 and v3 was **wrong**: since 2.4GHz appeared to be working, I just tried hardcoding MT7613 calibration data and that worked on one of my v3 devices. I later confirmed that 2.4GHz was not working fine, and since there is a much better solution, I closed that pull request.

My disussion with @Djfe on that PR and on the forum led me to this much better fix. I waited to submit it until I also had a v2 device to test with.

zekica commented:
What confused most people (including me) is that in the current OpenWrt .dts for v2/v3 the wifi chips are listed in the reverse order:

```
&pcie1 {
	wifi@0,0 {
...

&pcie0 {
	wifi@0,0 {
```

DragonBluep commented:
@zekica Hi! Could you please paste the v2/v3 `dmesg | grep -i pci` outputs here?

amhairghin's V2 [boot log](https://forum.openwrt.org/t/new-version-of-cudy-wr1300-v2-0/135490/15) shows 
```
[   11.454416] mt7603e 0000:02:00.0: ASIC revision: 76030010
```
And NPeca75's V3 [boot log](https://forum.openwrt.org/t/cudy-wr1300-v3-mt7613be/168743/4) shows
```
[    8.116596] mt7603e 0000:01:00.0: ASIC revision: 76030010
```

As you can see, they have different PCIe addresses. So I think the current eeprom offsets for V2 are correct (It's still need the eeprom size fix).

zekica commented:
@DragonBluep You are correct - v2 is running in production 22.03.x and I can't flash it right now, but here is the bootlog from the last boot:
https://gist.github.com/zekica/067037325c0cc3552e695395e776bee6

my v3:
https://gist.github.com/zekica/4424c201bcd8129cb964c05a144ba24e


From this, you are correct: so from my point of view, we need a new .dts for v3 (like in my old PR).

DragonBluep commented:
> From this, you are correct: so from my point of view, we need a new .dts for v3 (like in my old PR).

Yes. Please only fix the eeprom size in this commit and create another one to introduce v3 support.


> What confused me was that Cudy apparently didn't switch the pci nodes in the .dts they share with their openwrt fork image and offers one image for v2 and v3. I guess they made a mistake after all.

If they use the MTK vendor driver, the eeprom offset is useless. I remember the real offset is hard coded in driver source files. 

And I see `ieee80211-freq-limit` is missing, please consider adding it in v3 support patch. Adding this property can help us to identify which is the 5GHz NIC and limit the frequency band.

zekica committed:
ramips: Cudy WR1300v2 fix mt7613 calibration data length

Since MT7613 is handled by MT7615 driver, and other devices using MT7615
have reg = <0x8000 0x4da8>; this needs updating or eeprom data fails to load.

Signed-off-by: Filip Milivojevic <zekica@gmail.com>

zekica commented:
@DragonBluep I have done this. I don't have a v2 to test right now, but this (v2 calibration data length) change is 100% correct even when looking just at the code.

I'll create another pull request introducing Cudy WR1300v3 support.

Thank you for all your help!

akanouras commented:
Can we get this merged in 23.05 branch as well?

zekica commented:
I can open a new PR later today... I have it already backported to 23.05

zekica commented:
I don't think this needs backporting as current 23.05 DTS doesn't use nvmem cells but mtd-eeprom.

The current dts is **not strictly correct** but by the fact that it loads the entire partiton in ram, the mt76 driver has all the data it needs.

What do you think is the best solution here?

akanouras commented:
My apologies, I hastily assumed it was required in 23.05 as well.

fragtion commented:
I have a v4.0, and successfully flashed it with v3.0 firmware (Cudy has provided the openwrt-ramips-mt7621-cudy_wr1300-v4-squashfs-flash.zip on their Google Drive).
What else is needed to fully support this new variant and have it listed in OpenWrt's hardware table? Please let me know if/how I can help with this in any way

zekica commented:
You can start by testing that actually everything works - including LED lights and buttons on the device.

You can also extract the flattened device tree from their proprietary OpenWrt builds.

I have done so by using the following commands in linux:

```
binwalk -e openwrt-ramips-mt7621-cudy_wr1300-v4-squashfs-flash.bin
cd _openwrt-ramips-mt7621-cudy_wr1300-v4-squashfs-flash.bin.extracted/
```

Then you will see a file named `50040` - this is the kernel image with include flattened device tree binary format. `binwalk -e 50040` will tell you the offset of the DTB. Then you can extract it with dd and convert to a text format using `dtc`:
```
dd if=50040 bs=1 skip=9483456 of=device.dtb
dtc -i dtb -O dts device.dtb -o device.dts
```

Do exactly the same with the latest Cudy's proprietary OpenWrt build, and you will get it's (Flattened) Device Tree Source, and you will get these two files: [cudy-wr1300-dts.zip](https://github.com/user-attachments/files/25662405/cudy-wr1300-dts.zip)

_(start of some background info than may help you)_

These two files contain the entire device specification built from their **actual source device tree** but also includes all the hardware nodes for the SoC itself (sourced from [mt7621.dtsi](https://github.com/torvalds/linux/blob/master/arch/mips/boot/dts/ralink/mt7621.dtsi) - from the actual version that was uses in their build process).

I would start by ignoring all the nodes that are present in the base dtsi file and looking at what exactly was added (see [v2v3 shared dtsi](https://github.com/openwrt/openwrt/blob/main/target/linux/ramips/dts/mt7621_cudy_wr1300-v2v3.dtsi) and the respective v2 and v3 dts files to see what is added/modified in their buid).

_(end background info)_

Doing a diff on these two dts files results in the following file: [dts-diff.patch](https://github.com/user-attachments/files/25663067/dts-diff.patch) which contains the following differences:

> compatible = "cudy,wr1300-v4\0mediatek,mt7621-soc";
> model = "Cudy WR1300 v4";

(this specifies the device name but has to contain `mediatek,mt7621-soc` as it otherwise won't boot. In the actual source file the "comtabile" has to be written as ( \0 is a separator between two values):

> compatible = "cudy,wr1300-v4", "mediatek,mt7621-soc";

Then we see that the reset and wps buttons are different:

>  reset {
>    \- gpios = <0x16 0x12 0x01>;
>    \+ gpios = <0x16 **0x0a** 0x01>;
> wps {
>    \- gpios = <0x16 0x07 0x01>;
>    \+ gpios = <0x16 **0x0c** 0x01>;

These have to be written in the actual source as (for reset and wps respectively):

> gpios = <&gpio **10** GPIO_ACTIVE_LOW>;
> gpios = <&gpio **12** GPIO_ACTIVE_LOW>;

Then there are some more leds that have to be added/removed.
